gpflow/base.py

```diff
@@ -225,125 +190,11 @@ def assign(
-        return self.shape
-
-    def _should_act_as_resource_variable(self):  # type unknown
-        # needed so that Parameters are correctly identified by TensorFlow's
-        # is_resource_variable() in resource_variable_ops.py
-        pass  # only checked by TensorFlow using hasattr()
-
-    @property
-    def handle(self):  # type unknown
-        return self._unconstrained.handle
-
-    def __repr__(self) -> str:
-        unconstrained = self.unconstrained_variable
-        constrained = self.read_value()
-        if tf.executing_eagerly():
-            info = (
-                f"unconstrained-shape={unconstrained.shape} "
-                f"unconstrained-value={unconstrained.numpy()} "
-                f"constrained-shape={constrained.shape} "
-                f"constrained-value={constrained.numpy()}"
-            )
-        else:
-            if unconstrained.shape == constrained.shape:
-                info = f"shape={constrained.shape}"
-            else:
-                info = (
-                    f"unconstrained-shape={unconstrained.shape} "
-                    f"constrained-shape={constrained.shape}"
-                )
-
-        return f"<gpflow.Parameter {self.name!r} dtype={self.dtype.name} {info}>"
-
-    # Below
-    # TensorFlow copy-paste code to make variable-like object to work
-
-    @classmethod
-    def _OverloadAllOperators(cls):  # pylint: disable=invalid-name
-        """Register overloads for all operators."""
-        for operator in tf.Tensor.OVERLOADABLE_OPERATORS:
-            cls._OverloadOperator(operator)
-        # For slicing, bind getitem differently than a tensor (use SliceHelperVar
-        # instead)
-        # pylint: disable=protected-access
-        setattr(cls, "__getitem__", array_ops._SliceHelperVar)
-
-    @classmethod
-    def _OverloadOperator(cls, operator):  # pylint: disable=invalid-name
-        """Defer an operator overload to `ops.Tensor`.
-
-        We pull the operator out of ops.Tensor dynamically to avoid ordering issues.
-
-        Args:
-            operator: string. The operator name.
-        """
-        tensor_oper = getattr(tf.Tensor, operator)
-
-        def _run_op(a, *args, **kwargs):
-            # pylint: disable=protected-access
-            return tensor_oper(a.read_value(), *args, **kwargs)
-
-        functools.update_wrapper(_run_op, tensor_oper)
-        setattr(cls, operator, _run_op)
-
-    # NOTE(mrry): This enables the Variable's overloaded "right" binary
-    # operators to run when the left operand is an ndarray, because it
-    # accords the Variable class higher priority than an ndarray, or a
-    # numpy matrix.
-    # TODO(mrry): Convert this to using numpy's __numpy_ufunc__
-    # mechanism, which allows more control over how Variables interact
-    # with ndarrays.
-    __array_priority__ = 100
-
-
-Parameter._OverloadAllOperators()
-tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
-
 
 def _cast_to_dtype(
```

Comment:
Didn't TF2.2 resolve the issue with cast - can we simply use tf.case now?

Reply:
we still support 2.1